quickwit-storage/src/split_creation.rs

```diff
@@ -0,0 +1,404 @@
+) -> io::Result<(usize, usize)> {
+    let get_start = blocks_ranges
+        .iter()
+        .position(|block_range| block_range.end > range.start)
+        .ok_or_else(|| io::Error::new(ErrorKind::InvalidInput, "range is out of Range"))?;
+    let get_end = blocks_ranges
+        .iter()
+        .position(|block_range| block_range.end >= range.end)
+        .ok_or_else(|| io::Error::new(ErrorKind::InvalidInput, "range is out of Range"))?;
+    Ok((get_start, get_end))
+}
+
+/// SplitOffsetComputer is used to bundle together multiple files into a single split file.
+#[derive(Default, Debug)]
+pub struct SplitPayloadBuilder {
+    metadata: BundleStorageFileOffsets,
+    current_offset: usize,
+}
+
+impl SplitPayloadBuilder {
+    /// Creates a new SplitOffsetComputer, to which files and the hotcache can be added.
+    pub fn new() -> Self {
+        SplitPayloadBuilder {
+            current_offset: 0,
+            metadata: Default::default(),
+        }
+    }
+
+    /// Adds the file to the bundle file.
+    ///
+    /// The hotcache needs to be the last file that is added, in order to be able to read
+    /// the hotcache and the metadata in one continous read.
+    pub fn add_file(&mut self, path: &Path) -> io::Result<()> {
+        let file = std::fs::metadata(path)?;
+        let file_range = self.current_offset as u64..self.current_offset as u64 + file.len() as u64;
+        self.current_offset += file.len() as usize;
+        self.metadata.files.insert(path.to_owned(), file_range);
+        Ok(())
+    }
+
+    /// Writes the bundle file offsets metadata at the end of the bundle file,
+    /// and returns the byte-range of this metadata information.
+    pub fn finalize(self, hotcache: &[u8]) -> io::Result<SplitPayload> {
+        // Build the footer.
+        let mut footer_bytes = vec![];
+        // Fix paths to be relative
+        let metadata_with_fixed_paths = self
+            .metadata
+            .files
+            .iter()
+            .map(|(path, range)| {
+                let file_name = PathBuf::from(path.file_name().ok_or_else(|| {
+                    io::Error::new(
+                        ErrorKind::InvalidInput,
+                        format!("could not extract file_name from path {:?}", path),
+                    )
+                })?);
+                Ok((file_name, range.start..range.end))
+            })
+            .collect::<Result<HashMap<_, _>, io::Error>>()?;
+
+        let metadata_json = serde_json::to_string(&BundleStorageFileOffsets {
+            files: metadata_with_fixed_paths,
+            files_and_data: iter::once((
+                PathBuf::from(HOTCACHE_FILENAME.to_string()),
+                OwnedBytes::new(hotcache.to_owned()),
+            ))
+            .collect(),
+        })?;
+
+        footer_bytes.write_all(metadata_json.as_bytes())?;
+        let metadata_json_len = metadata_json.len() as u64;
+        footer_bytes.write_all(&metadata_json_len.to_le_bytes())?;
```

Comment:
`.extend`...

(The difference is mostly that we explicitly know we never return an error.)